pretyping/pretyping.ml

```diff
@@ -685,75 +677,84 @@ let rec pretype ~program_mode ~poly resolve_tc (tycon : type_constraint) (env :
           else fun f v -> applist (f, [v])
       | _ -> fun _ f v -> applist (f, [v])
     in
-    let rec apply_rec env sigma n resj candargs bidiargs = function
-      | [] -> sigma, resj, List.rev bidiargs
+    let nargs_before_bidi =
+      (* if `f` is a global, we retrieve bidirectionality hints *)
+      try
+        let (gr,_) = destRef sigma fj.uj_val in
+        Option.default length @@ GlobRef.Map.find_opt gr !bidi_hints
+      with DestKO ->
+        length
+    in
+    (* first_run: if true then candargs are allowed to fail unification and we do bidi *)
+    let rec apply_rec ~first_run env sigma n resj candargs restart = function
+      | [] -> sigma, resj, restart
       | c::rest ->
-        let bidi = n >= nargs_before_bidi in
+        let bidi = first_run && n >= nargs_before_bidi in
         let argloc = loc_of_glob_constr c in
         let sigma, resj = Coercion.inh_app_fun ~program_mode resolve_tc !!env sigma resj in
         let resty = whd_all !!env sigma resj.uj_type in
         match EConstr.kind sigma resty with
         | Prod (na,c1,c2) ->
           let tycon = Some c1 in
-          let (sigma, hj), bidiargs =
+          let (sigma, hj), restart =
             if bidi && Option.has_some tycon then
               (* We want to get some typing information from the context before
               typing the argument, so we replace it by an existential
               variable *)
               let sigma, c_hole = new_evar env sigma ~src:(loc,Evar_kinds.InternalHole) c1 in
-              (sigma, make_judge c_hole c1), (c_hole, c) :: bidiargs
-            else pretype tycon env sigma c, bidiargs
+              let rj, rn, rargs, rholes = Option.default (resj,n,[],[]) restart in
+              (sigma, make_judge c_hole c1), Some (rj,rn,c::rargs,c_hole::rholes)
+            else (assert (Option.is_empty restart); pretype tycon env sigma c, None)
           in
           let sigma, candargs, ujval =
             match candargs with
             | [] -> sigma, [], j_val hj
             | arg :: args ->
               begin match Evarconv.unify_delay !!env sigma (j_val hj) arg with
-                | exception Evarconv.UnableToUnify _ ->
-                  sigma, [], j_val hj
+                | exception Evarconv.UnableToUnify (sigma',err) ->
+                  if first_run then sigma, [], j_val hj
```

Comment:
Can you explain this>

Reply:
In the first run `candargs` is some program mode thing which is allowed to not unify. In the second run it's the holes produced for the args after bidi starts.